tensorflow/core/kernels/generate_box_proposals_op.cu.cc

```diff
@@ -0,0 +1,605 @@
+      dev_boxes_keep_flags));
+  ResetTensor<int8>(dev_boxes_keep_flags, d);
+  return Status::OK();
+}
+
+// Allocate workspace for NMS operation
+Status AllocatePreNMSTempTensors(
+    OpKernelContext* context, Tensor* dev_image_prenms_boxes,
+    Tensor* dev_image_prenms_scores, Tensor* dev_image_boxes_keep_list,
+    Tensor* dev_postnms_rois, Tensor* dev_postnms_rois_probs,
+    Tensor* dev_prenms_nboxes, int num_images, int num_boxes_to_generate,
+    int box_dim, int post_nms_topn, int pre_nms_topn) {
+  auto d = context->eigen_gpu_device();
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT, TensorShape({box_dim * num_boxes_to_generate}),
+      dev_image_prenms_boxes));
+  ResetTensor<float>(dev_image_prenms_boxes, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT, TensorShape({num_boxes_to_generate}),
+      dev_image_prenms_scores));
+  ResetTensor<float>(dev_image_prenms_scores, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_INT32, TensorShape({num_boxes_to_generate}),
+      dev_image_boxes_keep_list));
+  ResetTensor<int32>(dev_image_boxes_keep_list, d);
+
+  const int max_postnms_nboxes = std::min(num_boxes_to_generate, post_nms_topn);
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT,
+      TensorShape({box_dim * num_images * max_postnms_nboxes}),
+      dev_postnms_rois));
+  ResetTensor<float>(dev_postnms_rois, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_FLOAT, TensorShape({num_images * max_postnms_nboxes}),
+      dev_postnms_rois_probs));
+  ResetTensor<float>(dev_postnms_rois_probs, d);
+
+  TF_RETURN_IF_ERROR(context->allocate_temp(
+      DataType::DT_INT32, TensorShape({num_images}), dev_prenms_nboxes));
+  ResetTensor<int32>(dev_prenms_nboxes, d);
+
+  return Status::OK();
+}
+
+// Initialize index and offset arrays.
+// num_images is the batch size, KA is the number of anchors
+__global__ void InitializeDataKernel(const Cuda2DLaunchConfig config,
+                                     int* d_image_offsets,
+                                     int* d_boxes_keys_iota) {
+  const int KA = config.virtual_thread_count.x;
+  const int num_images = config.virtual_thread_count.y;
+  // printf("num_images %d KA %d\n",num_images,KA);
+  CUDA_AXIS_KERNEL_LOOP(img_idx, config.virtual_thread_count.y, Y) {
+    CUDA_AXIS_KERNEL_LOOP(box_idx, config.virtual_thread_count.x, X) {
+      // CUDA_2D_KERNEL_LOOP(box_idx, KA, img_idx, num_images) {
+      d_boxes_keys_iota[img_idx * KA + box_idx] = box_idx;
+
+      // One 1D line sets the 1D data
+      if (box_idx == 0) {
+        d_image_offsets[img_idx] = KA * img_idx;
+        // One thread sets the last+1 offset
+        if (img_idx == 0) d_image_offsets[num_images] = KA * num_images;
+      }
+    }
+  }
+}
+
+}  // namespace
+
+class GenerateBoundingBoxProposals : public tensorflow::AsyncOpKernel {
```

Comment:
It was to prevent stream synchronization needed with nms calls, however current event processing infrastructure of TF is insufficient for doing such kind of asynchronous processing thus idea was abandoned. I reverted it back to sync op.